core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,307 @@
+        let miniblock_number = connection
+            .blocks_dal()
+            .get_last_miniblock_for_l1_batch(&prev_l1_batch_number)
+            .await
+            .context(format!(
+                "get_last_miniblock_for_l1_batch({prev_l1_batch_number:?})"
+            ))?
+            .unwrap_or_else(|| {
+                panic!(
+                    "l1_batch_number {:?} must have a previous miniblock to start from",
+                    l1_batch_number
+                )
+            });
+
+        let fee_account_addr = connection
+            .blocks_dal()
+            .get_fee_address_for_l1_batch(&l1_batch_number)
+            .await?
+            .unwrap_or_else(|| {
+                panic!(
+                    "l1_batch_number {:?} must have fee_address_account",
+                    l1_batch_number
+                )
+            });
+        let (system_env, l1_batch_env) = load_l1_batch_params(
+            &mut connection,
+            l1_batch_number,
+            fee_account_addr,
+            validation_computational_gas_limit,
+            l2_chain_id,
+        )
+        .await
+        .unwrap();
+
+        drop(connection);
+
+        let connection_pool = connection_pool.clone();
+        // This spawn_blocking is needed given PostgresStorage's interface is a mix of blocking and non-blocking
+        // The interface may be either refactored or broken down in a async and sync interface.
+        tokio::task::spawn_blocking(move || {
+            let rt_handle = tokio::runtime::Handle::current();
+            let connection = rt_handle
+                .block_on(connection_pool.access_storage())
+                .unwrap();
+            let pg_storage =
+                PostgresStorage::new(rt_handle.clone(), connection, miniblock_number, true)
+                    .with_caches(PostgresStorageCaches::new(
+                        128 * 1_024 * 1_024, // 128MB -- picked as the default value used in EN
+                        128 * 1_024 * 1_024, // 128MB -- picked as the default value used in EN
+                    ));
+            let storage_view = StorageView::new(pg_storage).to_rc_ptr();
+
+            let mut vm = VmInstance::new(l1_batch_env, system_env, storage_view.clone());
+
+            let mut connection = rt_handle
+                .block_on(connection_pool.access_storage())
+                .unwrap();
+            let miniblock_and_transactions = rt_handle.block_on(
+                connection
+                    .transactions_dal()
+                    .get_miniblock_with_transactions_for_l1_batch(l1_batch_number),
+            );
+            tracing::info!("Started execution of l1_batch: {l1_batch_number:?}");
+            for (miniblock, txs) in miniblock_and_transactions {
+                tracing::debug!("Started execution of miniblock: {miniblock:?}");
+                for tx in txs {
+                    execute_tx(&tx, &mut vm);
```

Comment:
nit: I would add transaction logging at `debug` level as well - will be useful when we'll just start debugging